scanners/sslyze.py

```diff
@@ -180,62 +180,91 @@ def parse_sslyze(xml):
 
-        leaf = chain.select_one("certificate[position=leaf]")
+        if (constructed_chain and (len(constructed_chain) > 0)):
+            highest_constructed = parse_cert(constructed_chain[-1])
+            issuer = cert_issuer_name(highest_constructed)
+            if issuer:
+                data['certs']['constructed_issuer'] = issuer
+            else:
+                data['certs']['constructed_issuer'] = "(None constructed)"
+
+        leaf = parse_cert(served_chain[0])
+        leaf_key = leaf.public_key()
 
-        # Key algorithm and length for leaf certificate only.
-        data['certs']['key_type'] = leaf.select_one("subjectPublicKeyInfo publicKeyAlgorithm").text
-        data['certs']['key_length'] = int(leaf.select_one("subjectPublicKeyInfo publicKeySize").text)
+        if hasattr(leaf_key, "key_size"):
+            data['certs']['key_length'] = leaf_key.key_size
+        elif hasattr(leaf_key, "curve"):
+            data['certs']['key_length'] = leaf_key.curve.key_size
+        else:
+            data['certs']['key_length'] = None
+
+        if leaf_key.__class__ == cryptography.hazmat.backends.openssl.rsa._RSAPublicKey:
+            leaf_key_type = "RSA"
+        elif leaf_key.__class__ == cryptography.hazmat.backends.openssl.dsa._DSAPublicKey:
+            leaf_key_type = "DSA"
+        elif leaf_key.__class__ == cryptography.hazmat.backends.openssl.ec._EllipticCurvePublicKey:
+            leaf_key_type = "ECDSA"
+        else:
+            leaf_key_type == str(leaf_key.__class__)
+
+        data['certs']['key_type'] = leaf_key_type
 
         # Signature of the leaf certificate only.
-        data['certs']['leaf_signature'] = leaf.select_one("signatureAlgorithm").text
+        data['certs']['leaf_signature'] = leaf.signature_hash_algorithm.name
 
         # Beginning and expiration dates of the leaf certificate
-        before_date = leaf.select_one("validity notBefore").text
-        after_date = leaf.select_one("validity notAfter").text
-        data['certs']['not_before'] = dateutil.parser.parse(before_date)
-        data['certs']['not_after'] = dateutil.parser.parse(after_date)
+        data['certs']['not_before'] = leaf.not_valid_before
+        data['certs']['not_after'] = leaf.not_valid_after
 
         # Look at only served leaf and intermediate certificates
-        signatures = target.select("certificateChain certificate[position=leaf],certificate[position=intermediate] signatureAlgorithm")
-        any_sha1 = False
-        for signature in signatures:
-            if "sha1With" in signature.text:
-                any_sha1 = True
-        data['certs']['any_sha1'] = any_sha1
+        data['certs']['any_sha1'] = target['certinfo']['has_sha1_in_certificate_chain']
 
     return data
+
+# Given the cert sub-obj from the sslyze JSON, use
+# the cryptography module to parse its PEM contents.
+def parse_cert(cert):
+    backend = cryptography.hazmat.backends.openssl.backend
```

Comment:
I'm locking it to the openssl backend, since right now I have to introspect on exact class name to distinguish between RSA, DSA, and ECDSA keys. I'm going to file an issue upstream to ask about a way out of that kludge.